src/algorithms/nspso.cpp

```diff
@@ -0,0 +1,526 @@
+                                  - 2 * d2 * d3 + d1 + d2 + d3)
+                        / (2 * (ndc_size - 1));
+            } else { // for higher dimension we just divide equally the entire volume containing the pareto front
+                for (decltype(nadir_point.size()) i = 0; i < nadir_point.size(); ++i) {
+                    delta *= nadir_point[i] - ideal_point[i];
+                }
+                delta = pow(delta, 1.0 / static_cast<double>(nadir_point.size()))
+                        / static_cast<double>(non_dom_chromosomes.size());
+            }
+
+            std::vector<vector_double::size_type> count(non_dom_chromosomes.size(), 0);
+            std::vector<vector_double::size_type> sort_list(non_dom_chromosomes.size());
+            compute_niche_count(count, non_dom_chromosomes, delta);
+            std::iota(std::begin(sort_list), std::end(sort_list), vector_double::size_type(0));
+            std::sort(
+                sort_list.begin(), sort_list.end(), [&count](decltype(count.size()) idx1, decltype(count.size()) idx2) {
+                    return detail::less_than_f(static_cast<double>(count[idx1]), static_cast<double>(count[idx2]));
+                });
+
+            if (ndf[0].size() > 1) {
+                for (decltype(sort_list.size()) i = 0; i < sort_list.size(); ++i) {
+                    best_non_dom_indices.push_back(ndf[0][sort_list[i]]);
+                }
+            } else { // ensure the non-dom population has at least 2 individuals (to avoid convergence to a point)
+                best_non_dom_indices.push_back(ndf[0][0]);
+                best_non_dom_indices.push_back(ndf[1][0]);
+            }
+        } else { // m_diversity_method == max min
+            vector_double maxmin(swarm_size, 0);
+            compute_maxmin(maxmin, fit);
+
+            std::iota(std::begin(sort_list_2), std::end(sort_list_2), vector_double::size_type(0));
+            std::sort(sort_list_2.begin(), sort_list_2.end(),
+                      [&maxmin](decltype(maxmin.size()) idx1, decltype(maxmin.size()) idx2) {
+                          return detail::less_than_f(maxmin[idx1], maxmin[idx2]);
+                      });
+            best_non_dom_indices = sort_list_2;
+            vector_double::size_type i;
+            for (i = 1u; i < best_non_dom_indices.size() && maxmin[best_non_dom_indices[i]] < 0; ++i)
+                ;
+            if (i < 2) {
+                i = 2; // ensure the non-dom population has at least 2 individuals (to avoid convergence to a point)
+            }
+
+            best_non_dom_indices = std::vector<vector_double::size_type>(
+                best_non_dom_indices.begin(),
+                best_non_dom_indices.begin()
+                    + static_cast<vector_double::difference_type>(i)); // keep just the non dominated
+        }
+
+        // Decrease w from m_max_w to m_min_w troughout the run
+        double w = m_max_w - (m_max_w - m_min_w) / m_gen * gen;
```

Comment:
@darioizzo @bluescarni 
From my understanding, due to this line, I cannot implement `memory` in this algorithm.. the problem is that when memory is inserted, there is no way for me to know apriori the total number of generations (`m_gen`), although I could count the `gen`.

Reply:
@sceki Indeed. These are shitty heuristic that I am surprised people still use. Defining a schedule for w is ok, but having it depend on m_gen/gen is horrible. Is this in the original paper too?

Reply:
@darioizzo  I actually do not think so. I believe that the authors have decided this arbitrarily. In one of the three papers that they mention as reference ( https://titan.csit.rmit.edu.au/~e46507/publications/nspso-gecco03.pdf ), it is said that `w` is gradually decreased from 1.0 to 0.4 .. not sure how they achieved this decrease (which probably also makes the results difficult to reproduce)

Reply:
Would it then make sense to leave it fixed? The user could then implement a schedule on ```w``` manually. something like:
```
pop = algo.evolve(pop)
algo.extract(nspso).set_w(2u)
```

Reply:
Sure. This seems more meaningful.

This could actually also be done in the normal pso, in which, right now, the weight is only passed as input ( and it is thus fixed )

Reply:
If w is fixed do you still need memory in this algorithm? Are there data members that change during the evolve?

Reply:
Sorry but right now I’m not responsive.. I’ll be operative again and take care of this from the 16th!

Reply:
ok .....

Reply:
> If w is fixed do you still need memory in this algorithm? Are there data members that change during the evolve?

If I understood well the question yes. There are other variables that would change when memory is implemented, and thus require a workaround if memory wants to be introduced (should not be difficult though). Shall I make `w` fixed and chosen by the user and also modify the algorithm in such a way that memory is introduced? 
Or shall I just do the first and leave the algo w/o memory?

Reply:
Make w fixed. If the algorithm has then still some state, introduce the memory kwarg as ausual.

Final objective is that (in pseudocode, api is not respected)
```pop = my_algo(gen=100).evolve(pop)```
and
```
for i in range(100)
    pop = my_algo(gen=1, memory=True).evolve(pop)
```